lib/modem_info/modem_info.c

```diff
@@ -0,0 +1,412 @@
+	[MODEM_INFO_OPERATOR] = "OPERATOR",
+	[MODEM_INFO_CELLID] = "CELLID",
+	[MODEM_INFO_IP_ADDRESS] = "IP ADDRESS",
+	[MODEM_INFO_UICC] = "SIM",
+	[MODEM_INFO_BATTERY] = "BATTERY",
+	[MODEM_INFO_TEMP] = "TEMP",
+	[MODEM_INFO_FW_VERSION] = "FW",
+};
+
+static rsrp_cb_t modem_info_rsrp_cb;
+
+static struct at_param_list m_param_list;
+static struct k_thread socket_thread;
+static K_THREAD_STACK_DEFINE(socket_thread_stack, THREAD_STACK_SIZE);
+static struct k_mutex socket_mutex;
+static int at_socket_fd = INVALID_DESCRIPTOR;
+static struct pollfd fds;
+static int nfds;
+
+static int at_cmd_send(int fd, const char *cmd, size_t size, char *resp_buffer)
+{
+	int len;
+
+	k_mutex_lock(&socket_mutex, K_FOREVER);
+
+	LOG_DBG("send: %s\n", cmd);
+	len = send(fd, cmd, size, 0);
+	LOG_DBG("len: %d\n", len);
+	if (len != size) {
+		LOG_DBG("send: failed\n");
+		return -EIO;
+	}
+
+	if (resp_buffer != NULL) {
+		len = recv(fd, resp_buffer, CONFIG_MODEM_INFO_BUFFER_SIZE, 0);
+	}
+
+	if (len < 0) {
+		return -EIO;
+	}
+
+	k_mutex_unlock(&socket_mutex);
+
+	return 0;
+}
+
+static bool is_cesq_notification(char *buf, size_t len)
+{
+	if ((len < CMD_SIZE(AT_CMD_CESQ_RESP)) ||
+	    (buf[4] != AT_CMD_CESQ_RESP[4])) {
+		return false;
+	}
+
+	return strstr(buf, AT_CMD_CESQ_RESP) ? true : false;
+}
+
+enum modem_data_type modem_info_type_get(enum modem_info info)
+{
+	__ASSERT(info < MODEM_INFO_COUNT, "Invalid argument.");
+
+	return modem_data[info]->data_type;
+}
+
+int modem_info_name_get(enum modem_info info, char *name)
+{
+	int len;
+
+	if (name == NULL) {
+		return -EINVAL;
+	}
+
+	len = strlen(modem_data_name[info]);
+
+	memcpy(name,
+		modem_data_name[info],
+		len);
+
+	return len <= 0 ? -EINVAL : len;
+}
+
+int modem_info_short_get(char *out, size_t param_index,
+			size_t len, size_t off)
+{
+	int err;
+	u16_t param_value;
+
+	err = at_params_get_short(&m_param_list, param_index,
+				  &param_value);
+
+	if (err != 0) {
+		return err;
+	}
+
+	return snprintf(out, len,
+			"%d", param_value - off);
+}
+
+int modem_info_string_get(char *buf, size_t param_index, size_t len)
+{
+	return at_params_get_string(&m_param_list, param_index,
+				    buf, len);
+}
+
+int modem_info_parse(const struct modem_info_data *modem_data, char *buf)
+{
+	int err;
+	u32_t param_index;
+
+	err = at_parser_max_params_from_str(
+		&buf[CMD_SIZE(modem_data->cmd)], &m_param_list,
+		modem_data->param_count);
+
+	if (err != 0) {
+		return err;
+	}
+
+	param_index = at_params_get_valid_count(&m_param_list);
+	if (param_index != modem_data->param_count) {
+		return -EAGAIN;
+	}
+
+	return err;
+}
+
+int modem_info_update(enum modem_info info, char *buf)
+{
+	int err;
+	int len = 0;
+	char recv_buf[CONFIG_MODEM_INFO_BUFFER_SIZE] = {0};
+
+	if (buf == NULL) {
+		return -EINVAL;
+	}
+
+	err = at_cmd_send(at_socket_fd,
+			modem_data[info]->cmd,
+			strlen(modem_data[info]->cmd),
+			recv_buf);
+
+	if (err) {
+		return err;
+	}
+
+	err = modem_info_parse(modem_data[info], recv_buf);
```

Comment:
How do you know that you actually received the response for the command you just issued and not for instance a CESQ notification or some other notification?

Reply:
That is currently the largest weakness of the modem informer. The only solution I've come up with as of now is what you are currently working on with your library. Is there another (temporary) solution, perhaps?

Reply:
After some thinking I've decided to check if the received message equals to an expected response of that command, I believe this is feasible.

Reply:
Since your project is addressing this problem have I decided to keep this as is to avoid overlapping.